GetAwaiter is not enough for generalized async return types

> Starting with C# 7.0, any type that has an accessible `GetAwaiter` method. The object returned by the `GetAwaiter` method must implement the [System.Runtime.CompilerServices.ICriticalNotifyCompletion](https://docs.microsoft.com/dotnet/api/system.runtime.compilerservices.icriticalnotifycompletion) interface.

The listed requirements may suffice for an `await` expression, but they are not enough for defining an `async` function; the compiler will report error CS1983 unless the type also has an [AsyncMethodBuilderAttribute](https://docs.microsoft.com/dotnet/api/system.runtime.compilerservices.asyncmethodbuilderattribute) attribute.

```C#
using System;
using System.Runtime.CompilerServices;
using System.Threading.Tasks;

public struct Awaiter : ICriticalNotifyCompletion
{
    public void OnCompleted(Action action) => throw new NotImplementedException();

    public void UnsafeOnCompleted(Action action) => throw new NotImplementedException();
}

public struct ReturnType
{
    public Awaiter GetAwaiter() => throw new NotImplementedException();
}

public class Demo
{
    public async ReturnType RunAsync()
    {
        await Task.Yield();
    }
}
```

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: d055727b-c97c-4d25-39b8-a4d3f21d0444
* Version Independent ID: c8a7c6f1-8ee1-5030-db12-3db65a968501
* Content: [Async Return Types (C#)](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/concepts/async/async-return-types#feedback)
* Content Source: [docs/csharp/programming-guide/concepts/async/async-return-types.md](https://github.com/dotnet/docs/blob/master/docs/csharp/programming-guide/concepts/async/async-return-types.md)
* Product: **dotnet-csharp**
* Technology: **csharp-async**
* GitHub Login: @BillWagner
* Microsoft Alias: **wiwagn**

--- a/docs/csharp/programming-guide/concepts/async/async-return-types.md
+++ b/docs/csharp/programming-guide/concepts/async/async-return-types.md
@@ -33,12 +33,7 @@ In the following example, the `WaitAndApologizeAsync` method doesn't contain a `
 
 :::code language="csharp" source="snippets/async-return-types/async-returns2.cs" id="TaskReturn":::
 
-`WaitAndApologizeAsync` is awaited by using an await statement instead of an await expression, similar to the calling statement for a synchronous void-returning method. The application of an await operator in this case doesn't produce a value. To clarify the terms statement and expression, refer to the table below:
-
-| Await kind | Example                                      | Type                                   |
-|------------|----------------------------------------------|----------------------------------------|
-| Statement  | `await SomeTaskMethodAsync()`                | <xref:System.Threading.Tasks.Task>     |
-| Expression | `T result = await SomeTaskMethodAsync<T>();` | <xref:System.Threading.Tasks.Task%601> |
+`WaitAndApologizeAsync` is awaited by using an await statement instead of an await expression, similar to the calling statement for a synchronous void-returning method. The application of an await operator in this case doesn't produce a value. When the right operand of an `await` is a <xref:System.Threading.Tasks.Task%601>, the `await` expression produces a result of `T`. When the right operand of an `await` is a <xref:System.Threading.Tasks.Task>, the `await` and its operand are a statement.
 
 You can separate the call to `WaitAndApologizeAsync` from the application of an await operator, as the following code shows. However, remember that a `Task` doesn't have a `Result` property, and that no value is produced when an await operator is applied to a `Task`.
 
@@ -79,7 +74,7 @@ The following example shows the behavior of an async event handler. In the examp
 
 ## Generalized async return types and ValueTask\<TResult\>
 
-Starting with C# 7.0, an async method can return any type that has an accessible `GetAwaiter` method.
+Starting with C# 7.0, an async method can return any type that has an accessible `GetAwaiter` method that returns an instance of an *awaiter type*. In addition, the type returned from the `GetAwaiter` method must have the <xref:System.Runtime.CompilerServices.AsyncMethodBuilderAttribute?displayProperty=nameWithType> attribute. You can learn more in the feature spec for [Task like return types](../../../../_csharplang/proposals/csharp-7.0/task-types.md).
 
 Because <xref:System.Threading.Tasks.Task> and <xref:System.Threading.Tasks.Task%601> are reference types, memory allocation in performance-critical paths, particularly when allocations occur in tight loops, can adversely affect performance. Support for generalized return types means that you can return a lightweight value type instead of a reference type to avoid additional memory allocations.
 

